### PR TITLE
Create and use AP::srv(), singleton for SRV_Channels

### PR DESCRIPTION
**Pull request by peterbarker**
Primarily to remove the (eg.) `blheli_ptr` hacks we have to make this class work kind-of-sort-of statically.

```
----------------------------  ---------  -----  ----------  ------  ----  ----------  -----  -----  ----
Board                         AP_Periph  blimp  bootloader  copter  heli  iofirmware  plane  rover  sub
3DRControlZeroG                          -72    *           -48     -72               -72    -64    -72
ACNS-CM4Pilot                            -8     *           16      -8                -8     0      -16
ACNS-F405AIO                             -8     *           16      0                 -8     0      -16
AIRLink                                  -72    *           -48     -64               -72    -64    -80
AR-F407SmartBat               *
ARKV6X                                   -72    *           -48     -64               -80    -64    -72
ARK_CANNODE                   -8
ARK_GPS                       24
ARK_RTK_GPS                   24
AeroFox-Airspeed              *
AeroFox-Airspeed-DLVR         *
AeroFox-GNSS_F9P              24
AeroFox-PMU                   -8
Airvolute-DCS2                           -64    *           -40     -56               -64    -56    -72
AnyleafH7                                -64    *           -48     -64               -64    -56    -64
Aocoda-RC-H743Dual                       -64    *           -48     -64               -64    -56    -64
AtomRCF405NAVI                           8      *           24      8                 8      8      0
BETAFPV-F405                                    *           24
BeastF7                                  0      *           24      8                 8      16     0
BeastF7v2                                8      *           24      0                 0      8      0
BeastH7                                  -72    *           -48     -64               -80    -64    -80
BeastH7v2                                -72    *           -48     -64               -80    -64    -80
BirdCANdy                     24
BlitzF745                                0      *           24      8                 0      8      0
BlitzF745AIO                             8      *           24      8                 0      8      0
BlitzH743Pro                             -64    *           -40     -64               -64    -56    -64
BlitzMiniF745                            0      *           24      8                 0      8      0
BlitzWingH743                            -64    *           -48     -64               -64    -56    -72
BotBloxDroneNet               *
C-RTK2-HP                     *
CBU-H7-Stamp                             -64    *           -40     -64               -64    -56    -64
CSKY405                                  8      *           24      8                 0      16     0
CUAV-7-Nano                              -72    *           -48     -64               -80    -64    -72
CUAV-Nora                                -72    *           -48     -64               -72    -64    -72
CUAV-Nora-bdshot                         -72    *           -48     -64               -80    -64    -80
CUAV-Pixhack-v3                          -64                -40     -64               -64    -56    -72
CUAV-X7                                  -72    *           -48     -64               -72    -64    -80
CUAV-X7-bdshot                           -72    *           -48     -64               -80    -64    -72
CUAV_GPS                      *
CUAVv5                                   -72    *           -48     -64               -72    -64    -72
CUAVv5-bdshot                            -72    *           -48     -64               -80    -64    -80
CUAVv5Nano                               -72    *           -48     -72               -80    -64    -72
CUAVv5Nano-bdshot                        -72    *           -48     -72               -80    -64    -80
CarbonixF405                  0
CarbonixL496                  0
CubeBlack                                -80    *           -56     -72               -88    -72    -88
CubeBlack+                               -80    *           -56     -80               -80    -72    -80
CubeBlack-periph              24
CubeGreen-solo                                  *           -56
CubeNode                      *
CubeNode-ETH                  *
CubeOrange                               -72    *           -48     -64               -72    -64    -72
CubeOrange-ODID                          -72    *           -48     -64               -80    -64    -72
CubeOrange-SimOnHardWare                                    -16
CubeOrange-bdshot                        -72    *           -48     -64               -72    -64    -72
CubeOrange-joey                          -72    *           -48     -72               -80    -64    -80
CubeOrange-periph             32
CubeOrange-periph-heavy       24
CubeOrangePlus                           -72    *           -48     -72               -80    -64    -80
CubeOrangePlus-SimOnHardWare                                -16
CubeOrangePlus-bdshot                    -72    *           -48     -64               -80    -64    -72
CubePurple                               -64    *           -48     -64               -64    -56    -72
CubeRedPrimary                           -64    *           -48     -64               -64    -56    -72
CubeRedPrimary-PPPGW          *
CubeRedSecondary                         -64    *           -40     -64               -64    -56    -72
CubeSolo                                        *           -56
CubeYellow                               -72    *           -48     -72               -80    -64    -72
CubeYellow-bdshot                        -72    *           -48     -72               -72    -64    -72
DevEBoxH7v2                              -8     *           8       -8                -8     0      -16
DrotekP3Pro                              -80    *           -56     -80               -88    -72    -80
Durandal                                 -72    *           -48     -64               -80    -64    -80
Durandal-bdshot                          -72    *           -48     -72               -72    -64    -72
F35Lightning                             0      *           24      0                 8      16     0
F4BY                                     -8     *           8       -8                -8     0      -16
FlyingMoonF407                           -8     *           8       -8                -8     0      -16
FlyingMoonF427                           -80    *           -56     -80               -80    -72    -80
FlyingMoonH743                           -72    *           -48     -72               -80    -64    -72
FlywooF405HD-AIOv2                              *           24
FlywooF405Pro                            0      *           24      0                 8      8      0
FlywooF405S-AIO                                 *           24
FlywooF745                               0      *           24      8                 0      8      0
FlywooF745Nano                           0      *           24      8                 0      8      0
FlywooH743Pro                            -64    *           -40     -64               -64    -56    -72
FoxeerF405v2                             8      *           24      0                 0      16     0
FoxeerH743v1                             -64    *           -48     -64               -64    -56    -72
FreeflyRTK                    24
G4-ESC                        24
GEPRCF745BTHD                            8      *           24      0                 0      8      0
H757I_EVAL                                      *
H757I_EVAL_intf                                 *
HEEWING-F405                                    *                                     8
HEEWING-F405v2                                  *                                     8
Here4AP                       24
Hitec-Airspeed                *
HitecMosaic                   *
HolybroF4_PMU                 *
HolybroG4_Airspeed            *
HolybroG4_Compass             *
HolybroG4_GPS                 24
HolybroGPS                    24
IFLIGHT_2RAW_H7                          -64    *           -40     -64               -64    -56    -64
JFB100                                   -72    *           -48     -64               -72    -64    -80
JFB110                                   -72    *           -48     -72               -80    -64    -80
JHEMCU-GSF405A                                  *           24
JHEMCU-GSF405A-RX2                              *           24
JHEMCU-H743HD                            -64    *           -40     -64               -64    -56    -64
JHEM_JHEF405                             -8     *           8       -8                -8     0      -8
KakuteF4                                 8      *           24      0                 8      8      0
KakuteF4-Wing                            0      *           24      0                 0      16     0
KakuteF4Mini                             0      *           24      8                 8      8      0
KakuteF7                                 8      *           24      0                 8      8      0
KakuteF7-bdshot                          8      *           24      8                 0      8      0
KakuteF7Mini                             8      *           24      8                 0      8      0
KakuteH7                                 -64    *           -40     -64               -64    -56    -72
KakuteH7-Wing                            -64    *           -48     -64               -64    -56    -64
KakuteH7-bdshot                          -64    *           -48     -64               -64    -56    -64
KakuteH7Mini                             -64    *           -48     -64               -64    -56    -64
KakuteH7Mini-Nand                        -64    *           -48     -64               -64    -56    -72
KakuteH7v2                               -64    *           -40     -64               -64    -56    -72
LongBowF405WING                          0      *           24      8                 0      16     0
MFT-SEMA100                              -64    *           -48     -64               -64    -56    -72
MUPilot                                  -72    *           -48     -64               -80    -64    -72
MambaF405-2022                           8      *           24      0                 0      8      0
MambaF405US-I2C                          0      *           24      8                 0      8      0
MambaF405v2                              0      *           24      8                 0      8      0
MambaH743v4                              -64    *           -40     -64               -64    -56    -72
MatekF405                                8      *           24      8                 0      8      0
MatekF405-CAN                            -8     *           8       -8                -8     0      -16
MatekF405-STD                            0      *           24      0                 8      16     0
MatekF405-TE                             8      *           24      0                 8      8      0
MatekF405-TE-bdshot                      0      *           24      0                 0      16     0
MatekF405-Wing                           8      *           24      8                 0      8      0
MatekF405-Wing-bdshot                    0      *           24      8                 8      16     0
MatekF405-bdshot                         0      *           24      8                 0      16     0
MatekF765-SE                             -64    *           -48     -64               -64    -56    -72
MatekF765-Wing                           -64    *           -40     -64               -64    -56    -72
MatekF765-Wing-bdshot                    -64    *           -40     -64               -64    -56    -64
MatekG474-DShot               0
MatekG474-Periph              *
MatekH743                                -64    *           -40     -64               -64    -56    -72
MatekH743-bdshot                         -64    *           -40     -64               -64    -56    -72
MatekH743-periph              24
MatekH7A3                                -64    *           -40     -64               -64    -56    -64
MatekL431-ADSB                *
MatekL431-APDTelem            24
MatekL431-Airspeed            *
MatekL431-BattMon             *
MatekL431-DShot               0
MatekL431-EFI                 *
MatekL431-GPS                 *
MatekL431-HWTelem             *
MatekL431-MagHiRes            *
MatekL431-Periph              24
MatekL431-Proximity           *
MatekL431-RC                  0
MatekL431-Rangefinder         *
MatekL431-Serial              *
MatekL431-bdshot              0
MazzyStarDrone                           -8     *           16      -8                -8     0      -16
MicoAir405Mini                           0      *           24      8                 0      8      0
MicoAir405v2                             0      *           24      8                 0      8      0
MicoAir743                               -64    *           -40     -64               -64    -56    -72
Nucleo-G491                   *
Nucleo-L476                   *
Nucleo-L496                   *
NucleoH743                                      *
NucleoH755                                      *
NxtPX4v2                                 -64    *           -48     -64               -64    -56    -64
OMNIBUSF7V2                              8      *           24      0                 8      16     0
OmnibusNanoV6                            0      *           24      8                 8      16     0
OmnibusNanoV6-bdshot                     0      *           24      0                 8      16     0
OrqaF405Pro                              8      *           24      8                 8      8      0
PH4-mini                                 -72    *           -48     -64               -72    -64    -72
PH4-mini-bdshot                          -72    *           -48     -72               -80    -64    -80
Pix32v5                                  -72    *           -48     -64               -72    -64    -72
PixC4-Jetson                             -72    *           -48     -64               -72    -64    -80
PixFlamingo                              -80    *           -56     -80               -88    -72    -80
PixFlamingo-F767                         -72    *           -48     -72               -80    -64    -80
PixPilot-C3                              -80    *           -56     -72               -88    -72    -88
PixPilot-V3                              -80    *           -56     -80               -80    -72    -80
PixPilot-V6                              -72    *           -48     -64               -72    -64    -72
PixPilot-V6PRO                           -72    *           -48     -72               -80    -64    -72
PixSurveyA1                              -64    *           -40     -64               -64    -56    -64
PixSurveyA1-IND                          -80    *           -56     -80               -80    -72    -80
PixSurveyA2                              -72    *           -48     -64               -72    -64    -72
Pixhawk1                                 -64    *           -40     -64               -64    -56    -72
Pixhawk1-1M                              -8     *           16      -8                -8     0      -16
Pixhawk1-1M-bdshot                       -8                 8       -8                -8     0      -16
Pixhawk1-bdshot                          -64                -40     -64               -64    -56    -64
Pixhawk4                                 -72    *           -48     -64               -80    -64    -80
Pixhawk4-bdshot                          -72    *           -48     -64               -72    -64    -80
Pixhawk5X                                -72    *           -48     -72               -72    -64    -80
Pixhawk6C                                -72    *           -48     -72               -80    -64    -80
Pixhawk6C-bdshot                         -72    *           -48     -72               -72    -64    -72
Pixhawk6X                                -72    *           -48     -72               -80    -64    -72
Pixhawk6X-ODID                           -72    *           -48     -72               -80    -64    -72
Pixhawk6X-PPPGW               *
Pixhawk6X-bdshot                         -72    *           -48     -64               -80    -64    -72
Pixracer                                 -80    *           -56     -80               -80    -72    -80
Pixracer-bdshot                          -80    *           -56     -80               -80    -72    -80
Pixracer-periph               24
QioTekAdeptF407                          -8     *           16      -8                -8     0      -16
QioTekZealotF427                         -80    *           -56     -80               -80    -72    -88
QioTekZealotH743                         -72    *           -48     -64               -72    -64    -72
QioTekZealotH743-bdshot                  -72    *           -48     -64               -80    -64    -80
R9Pilot                                  -64    *           -40     -64               -64    -56    -64
RADIX2HD                                 -16                16      -8                -16    0      -16
RadiolinkPIX6                            -64    *           -48     -64               -64    -56    -72
ReaperF745                               0      *           24      8                 0      8      0
SDMODELH7V1                              -64    *           -40     -64               -64    -56    -72
SDMODELH7V2                              -64    *           -48     -64               -64    -56    -72
SITL_arm_linux_gnueabihf                 -224               -144    -80               -144   -192   -112
SITL_x86_64_linux_gnu                    -184               -184    -184              -184   -184   -184
SIYI_N7                                  -72    *           -48     -72               -80    -64    -80
SPRacingH7                               -24    *           0       -24               -32    -16    -24
SPRacingH7RF                             -8     *           16      0                 -8     0      -8
Sierra-F405                   24
Sierra-F412                   24
Sierra-F9P                    24
Sierra-L431                   *
Sierra-PrecisionPoint         24
Sierra-TrueNavIC              *
Sierra-TrueNavPro             24
Sierra-TrueNavPro-G4          *
Sierra-TrueNorth              *
Sierra-TrueSpeed              *
SkySakuraH743                            -64    *           -48     -64               -64    -56    -64
SkystarsH7HD                             -72    *           -48     -72               -80    -64    -72
SkystarsH7HD-bdshot                      -72    *           -48     -64               -72    -64    -80
SpeedyBeeF405Mini                        0      *           24      0                 8      16     0
SpeedyBeeF405WING                        0      *           24      0                 0      8      0
SuccexF4                                 0      *           24      0                 8      16     0
Swan-K1                                         *                                     -72
TBS-Colibri-F7                           -72    *           -48     -72               -80    -64    -80
TBS_LUCID_H7                             -64    *           -48     -64               -64    -56    -72
TMotorH743                               -64    *           -48     -64               -64    -56    -72
VRBrain-v51                              -8     *           16      -8                -8     0      -16
VRBrain-v52                              -8     *           8       -8                -8     0      -8
VRBrain-v54                              -64    *           -40     -64               -64    -56    -64
VRCore-v10                               -8     *           16      -8                -8     0      -16
VRUBrain-v51                             -8     *           16      -8                -8     0      -8
VUAV-V7pro                               -72    *           -48     -64               -80    -64    -72
X-MAV-AP-H743v2                          -64    *           -40     -64               -64    -56    -64
YJUAV_A6                                 -72    *           -48     -72               -72    -64    -72
YJUAV_A6SE                               -16    *           16      -8                -16    0      -16
YJUAV_A6SE_H743                          -72    *           -48     -72               -72    -64    -72
YJUAV_A6Ultra                            -72    *           -48     -72               -72    -64    -72
ZeroOneX6                                -72    *           -48     -64               -72    -64    -72
ZubaxGNSS                     *
airbotf4                                 -8     *           16      -8                -8     0      -8
bbbmini                                  -192               -192    -192              -192   -192   -192
bebop                                    -192               -224    -160              -192   -224   -192
blue                                     -192               -240    -192              -192   -192   -240
canzero                                  -240               -192    -128              -208   -240   -224
crazyflie2                                      *           8
edge                                     -224               -224    -128              -192   -192   -272
erlebrain2                               -144               -192    -160              -240   -192   -192
f103-ADSB                     *
f103-Airspeed                 *
f103-GPS                      *
f103-HWESC                    *
f103-QiotekPeriph             *
f103-RangeFinder              *
f103-Trigger                  *
f303-GPS                      *
f303-HWESC                    *
f303-M10025                   *
f303-M10070                   *
f303-MatekGPS                 *
f303-PWM                      24
f303-TempSensor               24
f303-Universal                *
f405-MatekAirspeed            24
f405-MatekGPS                 24
fmuv2                                    -8                 16      -8                -8     0      -16
fmuv3                                    -64    *           -48     -64               -64    -56    -64
fmuv3-bdshot                             -64                -48     -64               -64    -56    -72
fmuv5                                    -72    *           -48     -64               -80    -64    -72
iomcu                                                                     *
iomcu-dshot                                                               *
iomcu-f103                                                                *
iomcu-f103-8MHz-dshot                                                     *
iomcu-f103-dshot                                                          *
iomcu_f103_8MHz                                                           *
kha_eth                       *
luminousbee4                             -96                -72     -88               -96    -88    -104
luminousbee5                             -72    *           -48     -72               -72    -64    -72
mRo-M10095                    *
mRoCANPWM-M10126              -8
mRoCZeroOEMH7-bdshot                     -72    *           -48     -64               -80    -64    -72
mRoControlZeroClassic                    -72    *           -48     -72               -72    -64    -72
mRoControlZeroF7                         -72    *           -48     -72               -72    -64    -72
mRoControlZeroH7                         -72    *           -48     -64               -80    -64    -80
mRoControlZeroH7-bdshot                  -72    *           -48     -64               -80    -64    -72
mRoControlZeroOEMH7                      -72    *           -48     -64               -72    -64    -80
mRoKitCANrevC                 *
mRoNexus                                 -72    *           -48     -72               -72    -64    -80
mRoPixracerPro                           -72    *           -48     -72               -80    -64    -72
mRoPixracerPro-bdshot                    -72    *           -48     -72               -72    -64    -72
mRoX21                                   -80    *           -56     -80               -80    -72    -80
mRoX21-777                               -72    *           -48     -72               -80    -64    -72
mindpx-v2                                -64    *           -40     -64               -64    -56    -72
mini-pix                                 -8     *           16      -8                -8     0      -8
modalai_fc-v1                            -72    *           -48     -72               -72    -64    -80
navigator                                -160               -192    -112              -192   -192   -192
navio                                    -144               -224    -160              -192   -192   -192
navio2                                   -192               -176    -192              -192   -192   -192
obal                                     -144               -176    -192              -192   -192   -144
omnibusf4                                8      *           24      8                 0      16     0
omnibusf4pro                             8      *           24      8                 0      16     0
omnibusf4pro-bdshot                      0      *           24      0                 0      16     0
omnibusf4pro-one                         0      *           24      0                 0      8      0
omnibusf4v6                              0      *           24      8                 0      8      0
pxf                                      -192               -240    -128              -192   -192   -240
pxfmini                                  -144               -192    -160              -240   -192   -192
rFCU                                            *
rGNSS                         24
revo-mini                                0      *           24      0                 8      16     0
revo-mini-bdshot                         8      *           24      8                 8      8      0
revo-mini-i2c                            8      *           24      0                 8      16     0
revo-mini-i2c-bdshot                     8      *           24      8                 8      16     0
revo-mini-sd                                    *
skyviper-f412-rev1                                          16
skyviper-journey                                            8
skyviper-v2450                                              40
sparky2                                  -8     *           16      -8                -8     0      -8
speedybeef4                              0      *           24      8                 8      16     0
speedybeef4v3                            0      *           24      8                 0      16     0
speedybeef4v4                            8      *           24      0                 8      16     0
sw-boom-f407                  24
sw-nav-f405                   24
sw-spar-f407                  24
thepeach-k1                              -64    *           -48     -64               -64    -56    -64
thepeach-r1                              -64    *           -48     -64               -64    -56    -64
----------------------------  ---------  -----  ----------  ------  ----  ----------  -----  -----  ----
```

### CODE REVIEW

**Comment by amilcarlucas** [comment]
Nice

**Comment by peterbarker** [comment]
I've flashed onto a CubeOrange and made sure servo outputs basically work.

File: libraries/SRV_Channel/SRV_Channels.cpp [hunk 1]
```diff
@@ -619,3 +579,12 @@ void SRV_Channels::set_emergency_stop(bool state) {
 #endif
     emergency_stop = state;
 }
+
+namespace AP {
```
Comment (tridge) [inline libraries/SRV_Channel/SRV_Channels.cpp:583]: we should see why we don't have this in the header

File: libraries/SRV_Channel/SRV_Channels.cpp [hunk 2]
```diff
@@ -589,7 +549,7 @@ void SRV_Channels::zero_rc_outputs()
     for (uint8_t i=0; i<NUM_SERVO_CHANNELS; i++) {
         hal.rcout->write(i, 0);
     }
-    push();
+    AP::srv().push();
```
Comment (tridge) [inline libraries/SRV_Channel/SRV_Channels.cpp:552]: should really match cork()

**Comment by peterbarker** [comment]
When you try to put the singleton-getter in the header:

```
SCB-WAF: /opt/gcc-arm-none-eabi-10-2020-q4-major/bin/../lib/gcc/arm-none-eabi/10.2.1/../../../../arm-none-eabi/bin/ld: lib/libArduCopter_libs.a(GPIO.cpp.3.o): in function `AP::srv()':GPIO.cpp:(.text._ZN2AP3srvEv+0x0): multiple definition of `AP::srv()'; ArduCopter/AP_Arming.cpp.50.o:AP_Arming.cpp:(.text._ZN2AP3srvEv+0x0): first defined here
SCB-WAF: GPIO.cpp:(.text._ZN2AP3srvEv+0x0): multiple definition of `AP::srv()'; ArduCopter/AP_Arming.cpp.50.o:AP_Arming.cpp:(.text._ZN2AP3srvEv+0x0): first defined herecollect2: error: ld returned 1 exit status
```

**Comment by andyp1per** [comment]
I think this is a good change. I would note that its quite a bad time in the release cycle to make such a cross-cutting change as it may make backports harder.

**Review by tridge — APPROVED** [review]

**Comment by andyp1per** [comment]
I think we should backport this to avoid annoyingly difficult integrations between branches

**Pull request merged**